server/grpc/rpc.go

```diff
@@ -56,28 +56,57 @@ func (s *RPC) Next(c context.Context, agentFilter rpc.Filter) (*rpc.Workflow, er
 		log.Debug().Msgf("agent connected: %s: polling", hostname)
 	}
 
-	fn := createFilterFunc(agentFilter)
+	filterFn := createFilterFunc(agentFilter)
 	for {
 		agent, err := s.getAgentFromContext(c)
 		if err != nil {
 			return nil, err
-		} else if agent.NoSchedule {
-			return nil, nil
 		}
 
-		task, err := s.queue.Poll(c, agent.ID, fn)
+		if agent.NoSchedule {
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		// poll blocks until a task is available
+		task, err := s.queue.Poll(c, agent.ID, filterFn)
 		if err != nil {
 			return nil, err
-		} else if task == nil {
+		}
+
+		if task == nil {
 			return nil, nil
 		}
 
+		agent, err = s.getAgentFromContext(c)
+		if err != nil {
+			return nil, err
+		}
+
+		// the agent is meanwhile marked as no-schedule, push the task back to the queue
+		if agent.NoSchedule {
+			// we need to mark the task as done, otherwise it will be stuck in the queue
+			if err := s.queue.Error(c, task.ID, errors.New("agent is marked as no-schedule")); err != nil {
+				log.Error().Err(err).Msgf("mark task '%s' error failed", task.ID)
+			}
+
+			task.AgentID = -1
+			task.ID += "1"
+			if err := s.queue.Push(c, task); err != nil {
+				log.Error().Err(err).Msgf("push task '%s' back to queue failed", task.ID)
+			}
+
+			time.Sleep(1 * time.Second)
+			continue
+		}
```

Comment:
Not sure if this is the best approach, or if we should drop all workers with a specific agentID from the agent update api call.

Reply:
changed it to use a context which seems much cleaner